pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/nativefst/FST.java

```diff
@@ -0,0 +1,438 @@
+  }
+
+  /**
+   * Visits all states reachable from <code>node</code> in postorder. Returning
+   * false from {@link StateVisitor#accept(int)} immediately terminates the
+   * traversal.
+   *
+   * @param v Visitor to receive traversal calls.
+   * @param <T> A subclass of {@link StateVisitor}.
+   * @param node Identifier of the node.
+   * @return Returns the argument (for access to anonymous class fields).
+   */
+  public <T extends StateVisitor> T visitInPostOrder(T v, int node) {
+    visitInPostOrder(v, node, new BitSet());
+    return v;
+  }
+
+  /** Private recursion. */
+  private boolean visitInPostOrder(StateVisitor v, int node, BitSet visited) {
+    if (visited.get(node)) {
+      return true;
+    }
+    visited.set(node);
+
+    for (int arc = getFirstArc(node); arc != 0; arc = getNextArc(arc)) {
+      if (!isArcTerminal(arc)) {
+        if (!visitInPostOrder(v, getEndNode(arc), visited)) {
+          return false;
+        }
+      }
+    }
+
+    return v.accept(node);
+  }
+
+  /**
+   * Same as {@link #visitInPreOrder(StateVisitor, int)}, starting from root
+   * automaton node.
+   *
+   * @param v Visitor to receive traversal calls.
+   * @param <T> A subclass of {@link StateVisitor}.
+   * @return Returns the argument (for access to anonymous class fields).
+   */
+  public <T extends StateVisitor> T visitInPreOrder(T v) {
+    return visitInPreOrder(v, getRootNode());
+  }
+
+  /**
+   * Visits all states in preorder. Returning false from
+   * {@link StateVisitor#accept(int)} skips traversal of all sub-states of a
+   * given state.
+   *
+   * @param v Visitor to receive traversal calls.
+   * @param <T> A subclass of {@link StateVisitor}.
+   * @param node Identifier of the node.
+   * @return Returns the argument (for access to anonymous class fields).
+   */
+  public <T extends StateVisitor> T visitInPreOrder(T v, int node) {
+    visitInPreOrder(v, node, new BitSet());
+    return v;
+  }
+
+  /**
+   * Build a map from a serialized string
+   *
+   * @param inputString Serialized string
+   * @return
+   */
+  protected Map<Integer, Integer> buildMap(String inputString) {
+    Map<Integer, Integer> hashMap = new HashMap<>();
+
+    inputString = inputString.substring(1, inputString.length() - 1);
+
+    if (inputString.isEmpty()) {
+      return hashMap;
+    }
+
+    String[] pairs = inputString.split(",");
+
+    for (String pair : pairs) {
+
+      String[] keyVal = pair.split("=");
+
+      int key = Integer.parseInt(keyVal[0].trim());
+      int val = Integer.parseInt(keyVal[1].trim());
+
+      hashMap.put(key, val);
+    }
+
+    return hashMap;
+  }
+
+  public abstract boolean isArcLast(int arc);
+
+  public void save(FileOutputStream fileOutputStream) {
+    try {
+      final byte[] fsaData =
+          new FSTSerializerImpl().withNumbers().serialize(this, new ByteArrayOutputStream()).toByteArray();
+
+      fileOutputStream.write(fsaData);
+    } catch (IOException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)